minidump-processor/src/symbols.rs

```diff
@@ -1,3 +1,98 @@
+//! This module defines the interface between minidump-processor and its [Symbolizer][].
+//!
+//! There can only be one [Symbolizer][], and this is configured by minidump-processor's Cargo
+//! feature flags. The currently defined Symbolizers are:
+//!
+//! * breakpad_symbols -- feature: breakpad-syms (currently the default)
+//! * symbolic -- feature: symbolic-syms (not yet implemented, but compiles)
+//!
+//!
+//! minidump-processor and the [Symbolizer][] communicate using a series of traits. The symbolizer
+//! must provide implementations of these traits:
+//!
+//! * [SymbolProvider][] - provides symbolication, cfi evaluation, and debug statistics
+//!     * Implemented by [Symbolizer][]
+//!     * This actually doesn't need to be a trait in the current design, it exists to allow
+//!       multiple symbolicators to be used together, via [MultiSymbolProvider][]. The other
+//!       SymbolProviders have been removed, but I figured it would be a waste to throw out
+//!       this minimally intrusive machinery.
+//!
+//! * [SymbolSupplier][] - maps a [Module][] to a [SymbolFile][]
+//!     * minidump-processor does not directly use this, it's just there so the Symbolizer can
```

Comment:
I think separating out "fetch the data necessary to symbolicate this module" from "use this symbol data to symbolicate this frame" is a useful distinction to maintain. @swatinem spelled out a strawman API proposal on Discord for what they might want from this crate which does maintain the distinction, I'll paste it here for context:

```rust
trait Module {
    type FrameIter: Iterator<Item = Frame>;

    /// Resolve the caller with CFI information
    fn walk_with_cfi(cpu_context: &CpuContext, stack_memory: Memory) -> CpuContext;

    /// Symbolicate Frames and Inlinees for an instruction address
    fn symbolicate_frame(cpu_context: CpuContext) -> FrameIter;
}

trait SymbolProvider {
    type M: Module;

    /// Resolves and loads a [`Module`] based on its [`ModuleId`]
    async fn resolve_module(module: &ModuleId) -> M;
}

struct Thread {
    initial_context: CpuContext,
    stacktrace: Vec<Frame>,
}

async fn minidump_process<P>(minidump: &Minidump, symbol_provider: P) -> Vec<Thread>
where
    P: SymbolProvider,
{}
```
I don't want to bikeshed but I do think that we should choose names that are not confusables for data structures within minidumps. :) The names already in the crate aren't great, I mostly cribbed them from Breakpad.